client/src/controllers/KeyboardController.test.js

```diff
@@ -0,0 +1,64 @@
+import { Application } from '@hotwired/stimulus';
+import { KeyboardController } from './KeyboardController';
+
+describe('KeyboardController', () => {
+  let app;
+  const buttonClickMock = jest.fn();
+
+  /**
+   * Simulates a keydown, keypress, and keyup event for the provided key.
+   */
+  const simulateKey = (
```

Comment:
> In KeyboardController.test.js the simulateKey function you wrote appears to dispatch all three events(keydown,keypress,keyup) when it's called. Does Mousetrap register a press as all three events combined in a sequence?

Great question @NXPY123 my suggestion for this approach was driven by two bits of information;

### 1. Mousetrap docs - https://craig.is/killing/mice#api.bind

> "It is recommended that you leave this argument out if you are unsure. Mousetrap will look at the keys you are binding and determine whether it should default to keypress or keydown."

Basically, the docs say 'we will work it out for you but it will be one of keypress or keydown'. So better to have our unit test simulator just fire them all and our unit test check the things that matter (that the underlying `click` was called).


### 2. Digging into Mousetrap tests

I basically just read through the unit tests within the Mousetrap library, it's a bit full on but I found the tests util file.

https://github.com/ccampbell/mousetrap/blob/master/tests/libs/key-event.js#L48

In this file we can see that the comments and the code indicate that it's running all three; keydown, keypress and then keyup (in that order).

Reply:
This is really helpful! I didn't think of looking through tests in the Mousetrap library before.